browser/browsing_data/brave_clear_browsing_data.cc

```diff
@@ -0,0 +1,188 @@
+  // Corresponds to "Content settings" checkbox in the Clear Browsing Data
+  // dialog.
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteSiteSettingsOnExit))
+    *remove_mask |=
+        ChromeBrowsingDataRemoverDelegate::DATA_TYPE_CONTENT_SETTINGS;
+
+  return (*remove_mask != 0);
+}
+
+// This method will, for each loaded profile that is not off-the-record, gather
+// the user specified types of data that need to be cleared. It will then get
+// the BrowsingDataRemover for that profile and call its RemoveAndReply method.
+// BrowsingDataRemover will create a number of tasks to clear the data. Because
+// these tasks, on their own, will neither prevent shutdown nor stop the profile
+// from being destroyed, we have to block shutdown execution from proceeding any
+// further. Otherwise the tasks will be cancelled and the profiles destroyed.
+// Since we can't actually block the UI thread, instead we implement the Wait
+// method below, which just runs a RunLoop. When a BrowsingDataRemover finishes
+// its tasks it will reply back to us by calling the OnBrowsingDataRemoverDone
+// method below. When that happens we decrement the counter of profiles that
+// need to be cleared. Once the counter reaches 0 we exit the RunLoop and let
+// shutdown proceed.
+void BrowsingDataRemovalWatcher::ClearBrowsingDataForLoadedProfiles(
+    BraveClearBrowsingData::OnExitTestingCallback* testing_callback) {
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  DCHECK(profile_manager);
+
+  std::vector<Profile*> profiles = profile_manager->GetLoadedProfiles();
+  for (Profile* profile : profiles) {
+    if (profile->IsOffTheRecord())
+      continue;
+    int remove_mask;
+    int origin_mask;
+    if (!GetClearBrowsingDataOnExitSettings(profile, &remove_mask,
+                                            &origin_mask))
+      continue;
+    ++num_profiles_to_clear_;
+    content::BrowsingDataRemover* remover =
+        content::BrowserContext::GetBrowsingDataRemover(profile);
+    observer_.Add(remover);
+    if (testing_callback)
+      testing_callback->BeforeClearOnExitRemoveData(remover, remove_mask,
+                                                    origin_mask);
+    remover->RemoveAndReply(base::Time(), base::Time::Max(), remove_mask,
+                            origin_mask, this);
+  }
+
+  Wait();
+}
+
+void BrowsingDataRemovalWatcher::Wait() {
+  if (num_profiles_to_clear_ > 0)
+    run_loop_.Run();
+}
+
+void BrowsingDataRemovalWatcher::OnBrowsingDataRemoverDone() {
+  --num_profiles_to_clear_;
+  if (num_profiles_to_clear_ > 0)
+    return;
+
+  run_loop_.Quit();
+}
+
+}  // namespace
+
+BraveClearBrowsingData::OnExitTestingCallback*
+    BraveClearBrowsingData::on_exit_testing_callback_ = nullptr;
+
+// static
+void BraveClearBrowsingData::ClearOnExit() {
+  TRACE_EVENT0("browser", "BraveClearBrowsingData::ClearOnExit");
+  // Do not clear browsing data when the OS is ending session (logout/reboot/
```

Comment:
Ideally, it would be nice to document this somewhere that's user-visible. If it's too much information for the UI, maybe a page on `support.brave.com`?

Reply:
@bsclifton, @rebron ^^

Reply:
support.brave.com might be the right audience for that- but we can definitely document custom trace events in our wiki.  @mkarolin I created a stub page, if you wanted to add that (and describe why you'd use it):
https://github.com/brave/brave-browser/wiki/Custom-tracing-events

Reply:
@bsclifton, I understood it that @fmarier was referring to the comment that we are not clearing browsing data on logout/reboot/shutdown.

Reply:
ah- LOL; that makes a lot more sense 😂 

cc'ing @Brave-Matt on that one - would be great to document this feature in general 😄 @Brave-Matt: this would be something we're planning for 0.63.x

Reply:
Sorry, what I was referring to was not the trace events, but the limitation that if you shutdown your computer, it won't clear your history.

Reply:
That is a very bad limitation, which would probably surprise a lot of people who use the feature. IMO, we should back this out and fix that problem.

Reply:
cc: @rebron for the above comment by @tomlowenthal ^^ 

We may want to compare with Muon- it's possible Muon had that same limitation also. If the current experience is WORSE than Muon (ex: Muon works as expected), then I definitely agree that we should consider hiding this new UI until that problem is resolved

Reply:
@tomlowenthal should https://github.com/brave/brave-browser/issues/3549 cover this use-case also?

Reply:
Yeah, that issue is intended to be the new partial fix for this… someday.